File: Design-Patterns-Using-Cpp/Factory/FactoryMethodAnimal.cpp
#include <iostream>

// Animal interface
class Animal {
public:
    virtual void voice() = 0;
    virtual ~Animal() {}  // Ensure a virtual destructor for proper cleanup
};

// Concrete Animal classes
class Dog : public Animal {
public:
    void voice() override {
        std::cout << "Bhow Bhow!!" << std::endl;
    }
};

class Cat : public Animal {
public:
    void voice() override {
        std::cout << "Meow Meow!!" << std::endl;
    }
};

// AnimalFactory interface
class AnimalFactory {
public:
    virtual Animal* getAnimal() = 0;
    virtual ~AnimalFactory() {}  // Ensure a virtual destructor for proper cleanup
};

// Concrete AnimalFactory classes
class CatFactory : public AnimalFactory {
public:
    Animal* getAnimal() override {
        return new Cat();
    }
};

class DogFactory : public AnimalFactory {
public:
    Animal* getAnimal() override {
        return new Dog();
    }
};

// Client code
int main() {
    AnimalFactory* dogFactory = new DogFactory();
    dogFactory->getAnimal()->voice();

    AnimalFactory* catFactory = new CatFactory();
    catFactory->getAnimal()->voice();

    // Future changes to include cow type of objects.
    class Cow : public Animal {
    public:
        void voice() override {
            std::cout << "Gooaa Gooaa!!" << std::endl;
        }
    };

    class CowFactory : public AnimalFactory {
    public:
        Animal* getAnimal() override {
            return new Cow();
        }
    };

    // Client code for Cow
    AnimalFactory* cowFactory = new CowFactory();
    cowFactory->getAnimal()->voice();

    // Cleanup
    delete dogFactory->getAnimal();
    delete catFactory->getAnimal();
    delete cowFactory->getAnimal();
    delete dogFactory;
    delete catFactory;
    delete cowFactory;

    return 0;
}

/*
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
*/

File: Design-Patterns-Using-Go/Factory/FactoryMethodAnimal.go
package main

import "fmt"

// Animal interface
type Animal interface {
	Voice()
}

// Dog type
type Dog struct{}

func (d *Dog) Voice() {
	fmt.Println("Bhow Bhow!!")
}

// Cat type
type Cat struct{}

func (c *Cat) Voice() {
	fmt.Println("Meow Meow!!")
}

// AnimalFactory interface
type AnimalFactory interface {
	GetAnimal() Animal
}

// CatFactory type
type CatFactory struct{}

func (cf *CatFactory) GetAnimal() Animal {
	return &Cat{}
}

// DogFactory type
type DogFactory struct{}

func (df *DogFactory) GetAnimal() Animal {
	return &Dog{}
}

// Cow type
type Cow struct{}

func (c *Cow) Voice() {
	fmt.Println("Gooaa Gooaa!!")
}

// CowFactory type
type CowFactory struct{}

func (cf *CowFactory) GetAnimal() Animal {
	return &Cow{}
}

// Client Code
func main() {
	// Dog
	dogFactory := &DogFactory{}
	dogFactory.GetAnimal().Voice()

	// Cat
	catFactory := &CatFactory{}
	catFactory.GetAnimal().Voice()

	// Cow
	cowFactory := &CowFactory{}
	cowFactory.GetAnimal().Voice()
}

/*
Bhow Bhow!!
Meow Meow!!
Gooaa Gooaa!!
*/
